trin-execution/tests/content_generation.rs

```diff
@@ -124,5 +124,6 @@ async fn test_we_can_generate_content_key_values_up_to_x() -> Result<()> {
         trin_execution.database.storage_cache.clear();
         println!("Block {block_number} finished. Total content key/value pairs: {content_pairs}");
     }
+    temp_directory.close()?;
```

Comment:
Yes, [but](https://docs.rs/tempdir/latest/tempdir/struct.TempDir.html)...

> It's important to ensure that handles (like [File](http://doc.rust-lang.org/std/fs/struct.File.html) and [ReadDir](http://doc.rust-lang.org/std/fs/struct.ReadDir.html)) to files inside the directory are dropped before the TempDir goes out of scope. The TempDir destructor will silently ignore any errors in deleting the directory; to instead handle errors call [TempDir::close](https://docs.rs/tempdir/latest/tempdir/struct.TempDir.html#method.close).

So calling `close()` helps us identify if we're doing anything to stop the directory deletion from working.

Also, I wonder if the way we exit trin will let the destructor work. I haven't tinkered in the exit code for a while, so it would be nice to double-check that the new ephemeral directory disappears, at least after a clean trin exit.

Reply:
> Also, I wonder if the way we exit trin will let the destructor work. I haven't tinkered in the exit code for a while, so it would be nice to double-check that the new ephemeral directory disappears, at least after a clean trin exit.

Currently (and after this pr), when running `trin` or `trin-execution` with "ephemeral" flag, the folder won't be deleted at exit. The way it works (at least in Linux) is that new directory will be created under `/tmp/trin/` (`/tmp/trin-execution`), that OS will delete during next boot.

I actually like this behavior as it allows me to inspect database manually, or even restart the same node (e.g. using `--data-dir`), but still count on that data being deleted by the OS on next restart.

The way this works is that you can call [`TempDir::into_path`](https://docs.rs/tempdir/latest/tempdir/struct.TempDir.html#method.into_path) that will destroy `TempDir` without deleting the directory. This is used both in `trin` and `trin-execution`. But also in couple of tests, where I think it shouldn't (it's easier to have utility function that will initialize the object you are testing, e.g. OverlayService, without having to keep track of TempDir).

> I guess it's good practice to explicitly close this, but just so I understand, all the explicity closing you're doing isn't necessary correct? Since once `temp_directory` goes out of scope, the directory will be deleted

What @carver said, plus doing it explicitly means that you can't accidental call  `into_path` on it (e.g. doing refactoring) that will result in not deleting it.

Reply:
Ah right, I remember this debate about ephemeral, now. Yes, I see the value in being able to inspect. I also am slightly annoyed that running tests spams my /tmp, especially when running something like `cargo watch -x 'test ...'` which can run quite a few tests. (and I reboot my laptop ... irregularly 😆 ) But if I'm forced to choose, the inspection currently wins out, on balance.

I guess we could find some way to alter this behavior for just tests, but I'm not sure what that solution would look like, and if you don't see a quick way to do that, then I don't want to hold up the PR for it.

Reply:
In my opinion, tests should delete their TempFile once they are done. But they need to be refactored to behave like that. I plan to look into this next week, during my :flamingo: week.

---

And for completeness, as I believe I didn't explain this in details in the previous post.
If you have TempDir and you need `PathBuf`, you can do it in two ways:

- `temp_dir.into_path()` - which will destroy `TempDir` and return `PathBuf ` that you can use
    - this will not delete the directory at all
- `temp_dir.path().to_path_buf()` - this returns `&Path` that is converted to `PathBuf`
    - this will delete temporary directory when `temp_dir` goes out of scope (or `temp_dir.close()` is called
    - in order for this to work correctly in test, you have to keep `temp_dir` as variable in the body of the test itself, otherwise directory might be deleted before test finishes (e.g. if temp_dir is used only in utility function that initializes everything)